### ISSUE TITLE
Unable To Register Crashpad WER DLL In Packaged Application

### ISSUE THREAD

**Issue by brianmichel** [getsentry/sentry-native#923]
**Description**

When running Sentry Native in a packaged Windows application I see a debug message that the SDK found my crashpad wer handler and registered it successfully. However, when I view `HKEY_CURRENT_USER\Software\Microsoft\Windows\Windows Error Reporting\RuntimeExceptionHelperModules` I do not see the dll in the list.

I believe this is because of registry virtualization in packaged application since when I turn that feature off I do successfully see the registry entry. 

Is there any advice on how to work around this problem without disabling registry virtualization?

**When does the problem happen**

- [ ] During build
- [X] During run-time
- [ ] When capturing a hard crash

**Environment**

- OS: Windows 11 Pro 10.0.22621
- Compiler: Microsoft (R) C/C++ Optimizing Compiler Version 19.36.32537 for x64
- CMake version and config: cmake version 3.28.20231108-gb690d81, RelWithDebInfo, SENTRY_BACKEND=crashpad

**Steps To Reproduce**

1. Build and run the Sentry Native SDK in a Packaged Windows application

**Log output**
```
[sentry] INFO using database path "C:\Users\brian\AppData\Local\Temp\io.sentry\<redacted>"
[sentry] DEBUG starting transport
[sentry] DEBUG starting background worker thread
[sentry] DEBUG starting backend
[sentry] DEBUG background worker thread started
[sentry] DEBUG starting crashpad backend with handler "C:\Users\brian\Development\arc\build\installer\arc-windows\crashpad_handler.exe"
[sentry] DEBUG using minidump URL "https://<redacted>/minidump/?sentry_client=sentry.native/0.6.6&sentry_key=<redacted>"
[sentry] DEBUG registering crashpad WER handler "C:\Users\brian\Development\arc\build\installer\arc-windows\crashpad_wer.dll"
[sentry] INFO started crashpad client handler
[sentry] DEBUG processing and pruning old runs
[sentry] DEBUG sending envelope
[sentry] DEBUG submitting task to background worker thread
[sentry] DEBUG executing task on worker thread
[sentry] DEBUG sending request using winhttp to "https://<redacted>.sentry.io:443/api/<redacted>/envelope/":
x-sentry-auth:Sentry sentry_key=<redacted>, sentry_version=7, sentry_client=sentry.native/0.6.6
content-type:application/x-sentry-envelope
content-length:389

[sentry] DEBUG received response:
HTTP/1.1 200 OK
Date: Thu, 07 Dec 2023 21:31:59 GMT
Via: 1.1 google
Content-Length: 2
Content-Type: application/json
Server: nginx
Vary: origin,access-control-request-method,access-control-request-headers
access-control-allow-origin: *
access-control-expose-headers: x-sentry-error,x-sentry-rate-limits,retry-after
cross-origin-resource-policy: cross-origin
x-envoy-upstream-service-time: 0
Strict-Transport-Security: max-age=31536000; includeSubDomains; preload
Alt-Svc: h3=":443"; ma=2592000,h3-29=":443"; ma=2592000
```

**Comment by supervacuus** [comment]
Hi @brianmichel, thanks for the report. 

I have yet to test this in a setup with a packaged app container. Before I dive into this, do you expect the fast-fail crashes not to be detected in your case (or have you verified it)? Because from what you are describing, the behavior seems consistent with registry virtualization (i.e., changes to `HKEY_CURRENT_USER` are written to a per-process/-user "private" location and won't be visible from outside). 

So, the only question is whether the WER subsystem has access to that "private" exception handler module list. If you already have proof that WER doesn't act on the app container's registry entry, my investigation would go in a different direction.

Sadly, while the [docs](https://learn.microsoft.com/en-us/windows/win32/api/werapi/nf-werapi-werregisterruntimeexceptionmodule#remarks) for `WerRegisterRuntimeExceptionModule()` say that the function requires us to register the `WER` module in the registry, it won't return an error-code if that wasn't the case (even without registry virtualization). So, we get no runtime feedback from the API on whether the registry setup is consistent with the expectations of the API registration. At least the write to the registry key was successful in your log. Otherwise, you would have gotten a warning.

**Comment by brianmichel** [comment]
Hey @supervacuus I do see `RaiseFailFastException` get caught and reported to Sentry correctly, but I guess the thing I'm worried about is that there are other classes of crashes which aren't getting caught since the handler is getting registered to the virtualized entry instead of where WER expected things to be. I'm pretty new to the error reporting system on Windows (and Windows development in general) so please let me know if you disagree here.

Without our work around that starts a new process to manually add the `crashpad_wer.dll` to `HKEY_CURRENT_USER\Software\Microsoft\Windows\Windows Error Reporting\RuntimeExceptionHelperModules` I do see an entry in `HKEY_USERS\<user id>\Software\Microsoft\Windows\Windows Error Reporting\RuntimeExceptionHelperModules` which would indeed point to virtualization working correctly. However, reading the WER documentation it's unclear to me if it's actually going to look into `HKEY_USERS` or if  the entry *must* be in the `HKCU` hive. The documentation about this functionality is lacking to say the best, https://learn.microsoft.com/en-us/windows/win32/api/werapi/nf-werapi-werregisterruntimeexceptionmodule but it does say:

> This function requires that the pwszOutOfProcessCallbackDll DLL be included in the [WER exception handler module list](https://learn.microsoft.com/en-us/windows/desktop/wer/wer-settings) in the registry.

Additionally it seems that the Sentry code is explicitly trying to write to the `HKEY_CURRENT_USER` not `HKEY_USERS`. Is there a place we can ask Microsoft about this kind of bug (sorry for the naive question, Windows dev noob!)

**Comment by brianmichel** [comment]
@supervacuus I'm also starting a discussion on the MS tech forumshttps://techcommunity.microsoft.com/t5/msix/windows-error-reporting-module-registration/m-p/4006922#M3919

**Comment by supervacuus** [comment]
> Hey @supervacuus I do see `RaiseFailFastException` get caught and reported to Sentry correctly, but I guess the thing I'm worried about is that there are other classes of crashes which aren't getting caught since the handler is getting registered to the virtualized entry instead of where WER expected things to be. I'm pretty new to the error reporting system on Windows (and Windows development in general) so please let me know if you disagree here.

The `WER` module catches only fast-fail crashes ([more details](https://github.com/getsentry/sentry-native/issues/875#issuecomment-1689564967)). If you get any fast-fail crash reported in this configuration, then it means that the WER module is working correctly. You can verify this in multiple ways, but the simplest one is to look at the sentry log when crashing because it will not log the messages coming from our `SEH` handler like

```
flushing session and queue before crashpad handler
```
and
```
handing control over to crashpad
```
but still report a crash. This is the sole purpose of the `WER` module.

> Without our work around that starts a new process to manually add the `crashpad_wer.dll` to `HKEY_CURRENT_USER\Software\Microsoft\Windows\Windows Error Reporting\RuntimeExceptionHelperModules` I do see an entry in `HKEY_USERS\<user id>\Software\Microsoft\Windows\Windows Error Reporting\RuntimeExceptionHelperModules` which would indeed point to virtualization working correctly. However, reading the WER documentation it's unclear to me if it's actually going to look into `HKEY_USERS` or if the entry _must_ be in the `HKCU` hive. The documentation about this functionality is lacking to say the best, https://learn.microsoft.com/en-us/windows/win32/api/werapi/nf-werapi-werregisterruntimeexceptionmodule but it does say:
> 
> > This function requires that the pwszOutOfProcessCallbackDll DLL be included in the [WER exception handler module list](https://learn.microsoft.com/en-us/windows/desktop/wer/wer-settings) in the registry.

Yes, the `WER` documentation doesn't explicitly explain the behavior in the registry virtualization context, which is not great. But as is often the case with the comprehensive Win32 docs, I think it means there is no special case.

> Additionally it seems that the Sentry code is explicitly trying to write to the `HKEY_CURRENT_USER` not `HKEY_USERS`. Is there a place we can ask Microsoft about this kind of bug (sorry for the naive question, Windows dev noob!)

No worries, these are super relevant questions for other users too: `HKEY_CURRENT_USER` always links to a concrete `HKEY_USERS`, or in the lingo of the MS registry, `HKEY_CURRENT_USER` is a sub key of `HKEY_USERS`.

Please don't hold back with any questions. I am also interested what Microsoft has to say. Maybe there are things missing in the docs.

**Comment by brianmichel** [comment]
@supervacuus so I did some testing, and I think I'm seeing the right behavior regarding the WER handler. When our application gets packaged I see entries into `HKU\..\..Windows Error Reporting` and if I crash our app with `RaiseFailFastException` I eventually see them trickle into Sentry. I think the confusing thing (or one of them) is that I guess Windows will _always_ put a crash dump into `%LocalAppData%\CrashDumps` which can make it seem like all the minidumps aren't getting 'caught' by Sentry.

I _think_ things are working as advertised (still waiting on some info from MS), but we can probably close this issue if you want. If anything I think better documentation around WER and Sentry's interplay might be nice. I know there's another issue asking for something similar so I don't think there need to be two tickets for the same time.

**Comment by supervacuus** [comment]
> If anything I think better documentation around WER and Sentry's interplay might be nice. I know there's another issue asking for something similar so I don't think there need to be two tickets for the same time.

Is there any information you think we should add to the response here: https://github.com/getsentry/sentry-native/issues/875#issuecomment-1689564967. I will create an "advanced" doc, where I will put detail information about the inner workings and interactions with OS facilities like `WER` handling and I will as first step take my response verbatim. Let me know if you have anything to add, thx!

**Comment by whiskhub** [comment]
FYI: for MSIX packaged apps witt AppXManifest files, there is now this supported way of registering the DLL handler using `desktop7:RuntimeExceptionHelperModule`: https://learn.microsoft.com/en-us/uwp/schemas/appxpackage/uapmanifestschema/element-desktop7-runtimeexceptionhelpermodule
